### PR TITLE
[OP] Topk and arange

### PR DESCRIPTION
**Pull request by sxjscience**
Continuing of https://github.com/dmlc/mxnet/pull/3952

1. Fix the argument of dot and batch_dot
2. New operators: topk, sort, argsort, arange. (Also, the previous internal operators _zeros and _ones have been added to the python script as mx.sym.zeros and mx.sym.ones)
3. Move operator to operator/tensor: BlockGrad
4. Add the argument dtype to initialization operators (_zeros, _ones, _arange)
5. Enable gradient for `argmax` and `argmin` by adding the gradient node to `_zeros`.

Some Notes:

The implementation of TopK is not so optimized at this stage. The implementation is to first sort the data and then keep the top k indices/values. Also, the "axis" argument is supported in topk and topk is equivalent to argsort and sort if we set "k<1" and ret_typ = "indices" or "value"

The arange operator has an additional "repeat" argument, which indicates the number of times each element will be repeated in the generation process.

### CODE REVIEW

**Comment by sxjscience** [comment]
@piiswrong One problem is that I use INT_MIN for the default values, which will be ugly in json.

**Comment by piiswrong** [comment]
@sxjscience see this https://github.com/dmlc/dmlc-core/pull/194.
You can use end=None directly now

File: python/mxnet/symbol.py [hunk 1]
```diff
@@ -1129,3 +1129,65 @@ def hypot(left, right):
...
+    shape :  int or sequence of ints
+        Shape of the new array.
+    dtype : type, optional
+        The value type of the NDArray, default to np.float32
+    Returns
+    -------
+    out : Symbol
+        The created Symbol
+    """
+    return _internal._zeros(shape=shape, dtype=dtype)
```
Comment (piiswrong) [inline python/mxnet/symbol.py]: do you still need this?

Comment (sxjscience) [inline python/mxnet/symbol.py]: @piiswrong No, we do not need this now. Previously I do this because the registered names are _zeros and _ones.

Comment (piiswrong) [inline python/mxnet/symbol.py]: we should _zeros and add an alias zeros

File: src/operator/tensor/init_op.h [hunk 2]
```diff
@@ -11,20 +11,89 @@
...
+struct RangeParam : public dmlc::Parameter<RangeParam> {
+  real_t start;
+  real_t stop;
+  real_t step;
+  int repeat;
+  std::string ctx;
+  int dtype;
+  DMLC_DECLARE_PARAMETER(RangeParam) {
+    DMLC_DECLARE_FIELD(start)
+    .set_default(std::numeric_limits<real_t>::infinity())
```
Comment (piiswrong) [inline src/operator/tensor/init_op.h]: Could you try dmlc::optional<int>? It has been merged. Just checkout dmlc-core master

**Pull request closed without merging**